standardization/src/main/scala/za/co/absa/enceladus/standardization/interpreter/stages/TypeParser.scala

```diff
@@ -15,185 +15,331 @@
+        val out =  fieldType.fields.map(ParserWorker(_, currentAttrPath, origSchema, Parent.struct(column)).standardize)
+        val cols = out.map(_.stdCol)
+        val errs = out.map(_.errors)
+        // condition for nullable error of the struct itself
+        val nullErrCond = column.isNull and lit(!field.nullable)
+        val dropChildrenErrsCond = column.isNull
+        // first remove all child errors if this is null
+        val errs1 = concat(
+          flatten(array(errs.map(x => when(dropChildrenErrsCond, typedLit(Seq[ErrorMessage]())).otherwise(x)): _*)),
+          // then add an error if this one is null
+          when(nullErrCond,
+            array(callUDF("stdNullErr", lit(currentAttrPath))))
+          .otherwise(
+            typedLit(Seq[ErrorMessage]())
+          )
+        )
+        // rebuild the struct
+        val str = struct(cols: _*).as(field.name)
+        ParseOutput(str, errs1)
+      }
+    }
+
+    trait PrimitivePW extends ParserWorker {
+      /** Defines the cast error logic for numeric and other primitive types **/
+      //the casting logic differs by class, but within is the same, no need to evalute more timez
+      private lazy val primitiveCastLogic: Column = getPrimitiveCastLogic
+      protected def getPrimitiveCastLogic: Column
+
+      def primitiveCastErrorLogic: Column = {
+        val castedCol = primitiveCastLogic
+        field.dataType match {
+          // here we also want to check the numeric overflow by comparing with the original value this could break with
+          // trailing or leading zeros
+          case _: IntegerType |
+               _: ShortType |
+               _: FloatType |
+               _: DoubleType |
+               _: DecimalType |
+               _: ByteType |
+               _: LongType => castedCol isNull //TODO actual overflow/underflow check (#251)
+          case _ => castedCol isNull
+        }
+      }
+
+      override def standardize()(implicit spark: SparkSession, udfLib: UDFLibrary): ParseOutput = {
+        val err = when((column isNull) and lit(!field.nullable),
+          array(callUDF("stdNullErr", lit(s"$currentAttrPath${if (isArrayElement) "[*]" else ""}")))
+        ).otherwise(
+          when((column isNotNull) and primitiveCastErrorLogic,
+            array(
+              callUDF("stdCastErr",
+                lit(s"$currentAttrPath${if (isArrayElement) "[*]" else ""}"),
+                column.cast(StringType))
+            )
+          ).otherwise(
+            typedLit(Seq[ErrorMessage]())
+          )
+        )
+
+        val std = when((size(err) === lit(0)) and (column isNotNull),
+          primitiveCastLogic
+        ).otherwise(
+          // scalastyle:off null
+          when(size(err) === lit(0), null
+          // scalastyle:on null
+          ).otherwise( Defaults.getDefaultValue(field) )
+        ) as field.name
+
+        ParseOutput(std, err)
+      }
+    }
+
+    trait SimplePW extends  PrimitivePW {
```

Comment:
I think you misunderstood my point, an `abstract` type extending a `sealed` type breaks the seal unless the `abstract` subtype is also sealed. Specifying `sealed` on concrete types a bit unconventional but I guess it could make sense in some situations. 

However, `sealed` specifies something that is being extended by a predefined set of subclasses. But extending a `case class` is generally speaking a[ bad idea](https://stackoverflow.com/questions/34561614/should-i-use-the-final-modifier-when-declaring-case-classes) (I would even like to enforce a `final case class` convention)

Reply:
Yeah, I went little over-zealous with the `sealed`. 😊